### PR TITLE
Initial flow implementation in mattermost-webapp

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Initial flow implementation in mattermost-webapp (With some small examples of usage)

#### Checklist
- [ ] Ran `make check-style` to check for style errors (required for all pull requests)
- [ ] Ran `make test` to ensure unit and component tests passed

### CODE REVIEW

File: components/form_error.jsx [hunk 1]
```diff
@@ -1,39 +1,35 @@
...
+export default class FormError extends React.Component<Props> {
+    static defaultProps = {
+        error: null,
+        errors: [],
     }
 
     render() {
-        if (!this.props.error && this.props.errors.length === 0) {
+        const {type, error, margin, errors} = this.props;
+        if (!error && (!errors || errors.length === 0)) {
```
Comment (cometkim) [inline components/form_error.jsx]:
```suggestion
        if (error || errors.length !== 0)) {
```

File: components/confirm_modal.jsx [hunk 2]
```diff
@@ -1,80 +1,28 @@
...
-         */
-        onExited: PropTypes.func,
-
-        /*
-         * Set to hide the cancel button
-         */
-        hideCancel: PropTypes.bool,
-    }
-
+type Props = {|
```
Comment (cometkim) [inline components/confirm_modal.jsx]:
```suggestion
type Props = {
    +show: boolean,
    +title: React.Node,
    +message: React.Node,
    +confirmButtonClass: string,
    +modalClass: string,
    +confirmButtonText: React.Node,
    +cancelButtonText?: React.Node,
    +showCheckbox?: boolean,
    +checkboxText: React.Node,
    +onConfirm: (boolean) => void,
    +onCancel: (boolean) => void,
    +onExited: () => void,
    +hideCancel?: boolean,
}
```

`+` means read-only property, and `void` is the type for `void(0)`(`undefined`)

File: components/status_icon.jsx [hunk 3]
```diff
@@ -13,13 +13,13 @@ import StatusOfflineIcon from 'components/svg/status_offline_icon';
...
-export default class StatusIcon extends React.PureComponent {
-    static propTypes = {
-        status: PropTypes.string,
-        className: PropTypes.string,
-        type: PropTypes.string,
-    };
+type Props = {|
+    status?: 'online' | 'away' | 'dnd' | 'offline',
+    className: string,
+    type: 'avatar' | null,
```
Comment (cometkim) [inline components/status_icon.jsx]:
```suggestion
    type: ?'avatar',
```

File: flow/file-loader.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// @flow
+const s: string = '';
+export default s;
```
Comment (cometkim) [inline flow/file-loader.js]:
```diff
-// @flow
-const s: string = '';
-export default s;
+declare export default string;
```

File: .flowconfig [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+[options]
+module.system.node.resolve_dirname=node_modules
+suppress_comment= \\(.\\|\n\\)*\\$FlowFixMe
+module.name_mapper='^actions\(.*\)$' -> '<PROJECT_ROOT>/actions/\1'
+module.name_mapper='^selectors\(.*\)$' -> '<PROJECT_ROOT>/selectors/\1'
+module.name_mapper='^reducers\(.*\)$' -> '<PROJECT_ROOT>/reducers/\1'
+module.name_mapper='^utils\(.*\)$' -> '<PROJECT_ROOT>/utils/\1'
+module.name_mapper='^components\(.*\)$' -> '<PROJECT_ROOT>/components/\1'
+module.name_mapper='.*\.\(mp3\|svg\|png\|jpg\|gif\)$' -> '<PROJECT_ROOT>/flow/file-loader.js'
```
Comment (cometkim) [inline .flowconfig]:
```suggestion
module.name_mapper.extension='mp3' -> '<PROJECT_ROOT>/flow/WebpackAsset.js.flow'
module.name_mapper.extension='svg' -> '<PROJECT_ROOT>/flow/WebpackAsset.js.flow'
module.name_mapper.extension='png' -> '<PROJECT_ROOT>/flow/WebpackAsset.js.flow'
module.name_mapper.extension='jpg' -> '<PROJECT_ROOT>/flow/WebpackAsset.js.flow'
module.name_mapper.extension='gif' -> '<PROJECT_ROOT>/flow/WebpackAsset.js.flow'
```

File: .flowconfig [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+[version]
+0.89.0
```
Comment (cometkim) [inline .flowconfig]: Need to update?

**Comment by jespino** [comment]
I going to close this until we decide to go for typescript or not.

**Comment by cometkim** [comment]
fair enough... other projects will have no problem using typescript, as long as the mattermost-redux project still uses flow, but provides additional typescript definitions.

**Pull request closed without merging**